### PR TITLE
Refactor out QKV in projection

### PR DESCRIPTION
**Pull request by erichan1**
Summary: Refactor to reduce amount of copied code for decoder by finding common chunks for encoder and decoder. QKV in projection is a reasonable unit to copy out.

Test Plan:
buck run mode/opt -c fbcode.platform=platform010 -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 //pytext/fb/tools:benchmark_transformers -- transformer --batch-size 64 --avg-sequence-length 235 --max-sequence-length 256 --iters 100  --module native

Benchmark and numerical tests work fine.

Reviewed By: mikekgfb

Differential Revision: D36138504

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -282,6 +282,50 @@ std::tuple<Tensor, Tensor, Tensor> transform_bias_rescale_qkv_cpu(
...
+      auto kv = gemm_nt(key, q_kv_weight_s[1]);
+      qkv = at::cat({q, kv}, 2);
+    }
+  } else {
+    auto q_k_v_weight_s = at::native::chunk(qkv_weight, 3, 0);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        q_k_v_weight_s.size() == 3,
+        "expected chunk to produce 3 tensors but it produced ",
+        q_k_v_weight_s.size());
+    // TODO: can we stay nested instead of using cat?
```
Comment (vadimkantorov) [inline aten/src/ATen/native/transformers/attention.cpp:319]: I think nested now triggers a cat internally anyway :( @cpuhrsch

File: aten/src/ATen/native/transformers/attention.cpp [hunk 2]
```diff
@@ -282,6 +282,50 @@ std::tuple<Tensor, Tensor, Tensor> transform_bias_rescale_qkv_cpu(
   return std::make_tuple(q_k_v_s[0], q_k_v_s[1], q_k_v_s[2]);
 }
 
+Tensor qkv_matmul(
```
Comment (jbschlosser) [inline aten/src/ATen/native/transformers/attention.cpp:285]: naming nit: something like `qkv_projection` is a bit more specific

**Review by jbschlosser — COMMENTED** [review]
LGTM! One minor nit, nothing blocking

File: aten/src/ATen/native/transformers/attention.cpp (hunk 1 above)
Comment (erichan1) [reply aten/src/ATen/native/transformers/attention.cpp:319]: Good to know. A later TODO.

**Review by jbschlosser — APPROVED** [review]

**Pull request closed without merging**